src/main/java/org/opentripplanner/middleware/controllers/api/ApiUserController.java

```diff
@@ -59,8 +201,46 @@ ApiUser preCreateHook(ApiUser user, Request req) {
      */
     @Override
     boolean preDeleteHook(ApiUser user, Request req) {
-        ApiGatewayUtils.deleteApiKeys(user.apiKeyIds);
+        ApiGatewayUtils.deleteApiKeys(user.apiKeys);
         return true;
     }
 
+    /**
+     * Get an Api user from Mongo DB based on the provided user id. If no user id is provided, attempt to get user from
+     * request auth token.
+     */
+    private static ApiUser getApiUser(Request req) {
+        Auth0UserProfile requestingUser = Auth0Connection.getUserFromRequest(req);
+        ApiUser user;
+        String userId = req.queryParamOrDefault("userId", null);
+        if (userId != null) {
+            // If user Id is provided, assumption is that the requester is an AdminUser, with permission to modify an API
+            // user.
+            if (!isUserAdmin(requestingUser)) {
+                logMessageAndHalt(req, HttpStatus.FORBIDDEN_403, "Must be an admin to perform this operation.");
+            }
+            user = Persistence.apiUsers.getById(userId);
+            if (user == null) {
+                logMessageAndHalt(
+                    req,
+                    HttpStatus.NOT_FOUND_404,
+                    String.format("No Api user matching the given user id (%s)", userId),
+                    null
+                );
+            }
+        } else {
+            // User Id is null, API user should be obtained from token.
+            user = requestingUser.apiUser;
+            if (user == null) {
+                logMessageAndHalt(
+                    req,
+                    HttpStatus.BAD_REQUEST_400,
+                    "A user id is required to create an api key",
```

Comment:
Suggestion: say rather: "The userId parameter is required."

Reply:
Updated.